pkg/events/step/reconciler.go

```diff
@@ -0,0 +1,135 @@
+)
+
+const (
+	annotationKey = "tom24d.plugin.step-observer/result"
+)
+
+// ReconcileStepEvent is entry point to reconcile taskrun to determine whether it should emit CloudEvent.
+func ReconcileStepEvent(ctx context.Context, taskrun *v1beta1.TaskRun, client v1beta1client.TaskRunInterface) error {
+	logger := logging.FromContext(ctx)
+	if taskrun.Status.TaskSpec == nil ||
+		len(GetSteps(taskrun)) < 1 {
+		logger.Infof("step events emission skipped as no step in taskrun: %s", taskrun.Name)
+		return nil
+	}
+
+	sent, err := initializeAnnotation(taskrun)
+	if err != nil {
+		logger.Fatalf("failed to initialize annotation: %v", err)
+		return nil
+	}
+	logger.Infof("RECEIVED annotation: %s, annotation in the run: %v", sent, taskrun.Annotations)
+
+
+	for i, step := range GetStepStatuses(taskrun) {
+		// start
+		if step.ContainerState.Running != nil {
+			// (Running) emit&mark started if i=0
+			if i == 0 && !taskrun.Status.StartTime.IsZero(){
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepStarted)
+			}
+		}
+		// emit&mark started if i!=0 && i-1 step marked as succeeded
+		if i != 0 && sent.IsMarked(GetStepStatuses(taskrun)[i-1].Name, resources.CloudEventTypeStepSucceeded) {
+			ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepStarted)
+		}
+
+		if step.ContainerState.Terminated != nil {
+			// skipped
+			// (Terminated) emit&mark skipped if i!=0 && i-1 step marked as failed||skipped,
+			// then continue to avoid being considered as failure
+			if i != 0 &&
+					(sent.IsMarked(GetStepStatuses(taskrun)[i-1].Name, resources.CloudEventTypeStepFailed) ||
+					sent.IsMarked(GetStepStatuses(taskrun)[i-1].Name, resources.CloudEventTypeStepSkipped)) {
+
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepSkipped)
+				continue
+			}
+			// succeeded
+			// (Terminated) emit&mark succeeded if exit code is 0
+			if step.Terminated.ExitCode == 0 {
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepSucceeded)
+			} else {
+				// failure
+				// (Terminated) emit&mark failed if exit code is not 0
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepFailed)
+			}
+
+		}
+	}
+
+	// TODO proper reconcileErr handling
+	patch, err := getPatch(sent)
+	if err != nil {
+		return err
+	}
+	logger.Infof("PATCH: %s", patch)
+	_, err = client.Patch(taskrun.Name, types.MergePatchType, patch)
+	if err != nil {
+		logger.Errorf("failed to PATCH taskrun: %v", err)
+	}
+	return nil
+}
+
+func ensureEventEmitted(ctx context.Context, annotation *resources.EmissionStatuses, stepName string, eventType resources.TektonPluginEventType) {
+	logger := logging.FromContext(ctx)
+
+	emissionStatus, err := annotation.GetStatus(stepName)
+	if err != nil {
+		logger.Fatalf("status of step: %s not found", stepName)
+		return
+	}
+
+	// no emission if already marked as sent
+	if !emissionStatus.IsMarked(eventType) {
```

Comment:
@vdemeester 
here it checks whether the eventType is already sent. If yes, skip it. If no, send event and mark it to the *resources.EmissionStatuses.